File: MaximumDifferenceBetweenNodeandAncestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: TreeNode) -> int:
        def findMaxMin(root,mx,mn):
            if(root==None): return mx-mn
            
            mx,mn=max(root.val,mx),min(root.val,mn) 
            left=findMaxMin(root.left,mx,mn) # max val subtree of root.left 
            right=findMaxMin(root.right,mx,mn)# max val subtree of root.right
            
            return max(left,right) 
            
        return findMaxMin(root,root.val,root.val)
        
        
        
            
        
                
                

File: MaximumDifferenceBetweenNodeandAncestor.cpp
/* Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans=0;
    int maxAncestorDiff(TreeNode* root) {
        findDiff(root,root->val,root->val);
        return ans;
    }
    
    void findDiff(TreeNode* node, int mx, int mn){ // |10-1| < |1-11| 
        if(!node) return ;
        
        mx=max(mx,node->val);
        mn=min(mn,node->val);
        ans=max(ans,mx-mn);
        
        findDiff(node->left,mx,mn);
        findDiff(node->right,mx,mn);
        
    }
};
